File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.py
def solution(maps):
    
    def dfs(x, y):
        stack = [(x, y)]
        days = 0
        
        while stack:
            cx, cy = stack.pop()
            
            if visited[cx][cy]:
                continue
            
            visited[cx][cy] = True
            days += int(maps[cx][cy])
            
            for dx, dy in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
                nx, ny = cx + dx, cy + dy
                
                if 0 <= nx < len(maps) and 0 <= ny < len(maps[0]) and not visited[nx][ny] and maps[nx][ny] != 'X':
                    stack.append((nx, ny))
                    
        return days
    
    answer = []
    
    r, c = len(maps), len(maps[0])
    visited = [[False]*c for _ in range(r)]
    
    for i in range(r):
        for j in range(c):
            if maps[i][j] != 'X' and not visited[i][j]:
                answer.append(dfs(i, j))
    
    if not answer:
        return [-1]
    
    answer.sort()
    
    return answer

File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.java
import java.util.*;

class Solution {
    int n, m;
    boolean[][] visited;
    int[] dx = {-1, 1, 0, 0};
    int[] dy = {0, 0, 1, -1};
    
    public int[] solution(String[] maps) {
        n = maps.length;
        m = maps[0].length();
        visited = new boolean[n][m];
        List<Integer> answer = new ArrayList<>();
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (!visited[i][j] && maps[i].charAt(j) != 'X') {
                    answer.add(bfs(i, j, maps));
                }
            }
        }
        
        if (answer.isEmpty()) return new int[]{-1};
        Collections.sort(answer);
        return answer.stream().mapToInt(Integer::intValue).toArray();
    }
    
    private int bfs(int sx, int sy, String[] maps) {
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{sx, sy});
        visited[sx][sy] = true;
        int sum = maps[sx].charAt(sy) - '0';
        
        while (!q.isEmpty()) {
            int[] now = q.poll();
            int x = now[0], y = now[1];
            for (int d = 0; d < 4; d++) {
                int nx = x + dx[d];
                int ny = y + dy[d];
                if (0 <= nx && nx < n && 0 <= ny && ny < m) {
                    if (!visited[nx][ny] && maps[nx].charAt(ny) != 'X') {
                        visited[nx][ny] = true;
                        sum += maps[nx].charAt(ny) - '0';
                        q.add(new int[]{nx, ny});
                    }
                }
            }
        }
        return sum;
    }
}